waveform_editor/dependency_graph.py

```diff
@@ -0,0 +1,65 @@
+from waveform_editor.derived_waveform import DerivedWaveform
+
+
+class DependencyGraph:
+    def __init__(self):
+        self.graph = {}
+
+    def add_node(self, name, dependencies):
+        self.graph[name] = set(dependencies)
+        try:
+            self.detect_cycles()
+        except RuntimeError as e:
+            del self.graph[name]
+            raise e
+
+    def remove_node(self, name):
+        if name in self.graph:
+            del self.graph[name]
+        for deps in self.graph.values():
+            deps.discard(name)
+
+    def rename_node(self, old_name, new_name, config):
+        dependents = {node for node, deps in self.graph.items() if old_name in deps}
```

Comment:
```suggestion
        dependents = [node for node, deps in self.graph.items() if old_name in deps]
```

Lists are more efficient to construct and loop over than sets.